tests/test_custom_boot_api.py

```
@@ -0,0 +1,137 @@
+
+    def test_extra_args_must_be_single_tokens(self) -> None:
+        with pytest.raises(ValueError, match="one kernel argument token"):
+            DirectKernelBoot(extra_args=("foo=bar baz=qux",))
+
+
+class TestEnsureBaseKernelForBackend:
+    """Kernel resolution hides published asset format details."""
+
+    @patch("smolvm.images.published.ensure_base_kernel")
+    def test_qemu_selects_image_kernel(self, mock_ensure: MagicMock) -> None:
+        kernel = Path("/tmp/vmlinux.image")
```

Ruff (0.15.15)

[error] Probable insecure usage of temporary file or directory: "/tmp/vmlinux.image"

(S108)